vendor/github.com/ethereum/go-ethereum/eth/downloader/downloader.go

```diff
@@ -527,6 +528,7 @@ func (d *Downloader) cancel() {
 // finish before returning.
 func (d *Downloader) Cancel() {
 	d.cancel()
+	go func() { d.cancelWg.Done() }() // Remove sentinel
```

Comment:
I am not exactly sure what's going on here. We call `Done` on a wait group in a goroutine and wait for that wait group just below...

Reply:
if nobody called Add - then Wait will just exit without blocking

Reply:
> is that WaitGroup will complain if Wait gets called and nobody called Add yet

That's not a case. You can call `Wait` on a `WaitGroup` without calling `Add` earlier: https://play.golang.org/p/Bc_mzGsiNAX

This looks like we can just get rid of `cancelWg` completely.

Reply:
as i understand this parts can be used from different goroutines, and this is where race comes from. i think the right solution is to synchronize such parts.

Reply:
it looks like that you need to check for cancel before adding to group that is waited below

```
diff --git i/vendor/github.com/ethereum/go-ethereum/eth/downloader/downloader.go w/vendor/github.com/ethereum/go-ethereum/eth/downloader/downloader.go
index eb64f05a5..fe64b0089 100644
--- i/vendor/github.com/ethereum/go-ethereum/eth/downloader/downloader.go
+++ w/vendor/github.com/ethereum/go-ethereum/eth/downloader/downloader.go
@@ -482,6 +482,11 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
 // spawnSync runs d.process and all given fetcher functions to completion in
 // separate goroutines, returning the first error that appears.
 func (d *Downloader) spawnSync(fetchers []func() error) error {
+       select {
+       case <-d.cancelCh:
+               return nil
+       default:
+       }
        errc := make(chan error, len(fetchers))
        d.cancelWg.Add(len(fetchers))
        for _, fn := range fetchers {
```

Reply:
> if nobody called Add - then Wait will just exit without blocking

@dshulyak, true except if you run it with `-race`, where maybe too simplistically it will think there was a race condition:

```
		// Increment waiters count.
		if atomic.CompareAndSwapUint64(statep, state, state+1) {
			if race.Enabled && w == 0 {
				// Wait must be synchronized with the first Add.
				// Need to model this is as a write to race with the read in Add.
				// As a consequence, can do the write only for the first waiter,
				// otherwise concurrent Waits will race with each other.
				race.Write(unsafe.Pointer(&wg.sema))
			}
			runtime_Semacquire(&wg.sema)
```

Reply:
it is because Add and Wait are called from different goroutines without synchronization

Reply:
@dshulyak you're right, that makes much more sense. However, I don't think we should return `nil` as the error, otherwise we won't return in https://github.com/status-im/status-go/blob/master-travis-matrix-tests/vendor/github.com/ethereum/go-ethereum/eth/sync.go#L199

Reply:
yes maybe, tests passed locally but i didn't follow the whole codepath.
also it might be not enough to just check cancelCh, it still possible that Wait will be called before Add, but the window is much narrower. ideally we lock then check a cancel channel and only then spawn fetchers and then unlock

Reply:
@dshulyak I've changed to your suggestion, and the build is passing. Please re-review.

Reply:
hmm, this seems to have more implications in some situations causing peers to not get dropped. See test upstream: https://travis-ci.org/ethereum/go-ethereum/jobs/372143026#L2737

Reply:
i applied fix locally and it passes ethereum tests. i think that is a flake